### DISCUSSION TITLE
Ability to convert between certain field types

### DISCUSSION THREAD

**Discussion by Mosnar** [verbb/formie#2110] [Ideas]
A client created some forms and instead of using the "email" field, they used the "plain text" field. This is problematic since it's resulting in validation errors with their CRM integration.

As far as I can tell, there isn't a way to convert the existing text field to an email field; so, I'd have to delete the existing field and re-add the new one. The problem with this approach is that it will result in data loss since submissions will lose the deleted field's content.

Currently, I'm suggesting the client hides the old field, marks it as unrequired, and adds a new one. This is messy, but it works in a pinch.

Ideally, there'd be a an easy way to convert between compatible field types or at the very least replace a field without losing old submission content.

Related: #1682

**Comment by engram-design** [comment]
Correct, you can't convert field types, mostly to alleviate complexity. It's mostly a matter of how to handle converting content without any loss. For example, imagine someone trying to convert a Single-Line Text field to an Address field, or in an extreme case, a Repeater field.

If we were to implement this, we'd need to add some sort of compatible categorisation of fields to maintain what can be converted to what. Probably a `SingleValueFieldInterface`, `MultiValueFieldInterface`, etc for example. I'd _really_ like to not get into a scenario where we're modifying the content of submissions when you decide to change the field type, as that could just lead to issues.

We'd also need to normalize field settings, as each field has different settings, some of which are incompatible.

But yes, I totally get that right now it's a bit of a hassle. Short of modifying the field's type in the database (and tweaking the `settings` column), there's not a way to do this yet.

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.5](https://github.com/verbb/formie/releases/tag/4.0.0-beta.5). For now, it's pretty limited between Single-Line Text to Email, and we'll likely expand, but we're not looking to convert between radically different fields. That's just going to give us support headaches and annoy users.
